python_etl/churn.py

```
@@ -0,0 +1,654 @@
+#
+# Run the aggregation code, detecting files that are missing.
+#
+# The fields we want in the output are:
+#  - channel (appUpdateChannel)
+#  - geo (bucketed into top 30 countries + "rest of world")
+#  - is_funnelcake (contains "-cck-"?)
+#  - acquisition_period (cohort_week)
+#  - start_version (effective version on profile creation date)
+#  - sync_usage ("no", "single" or "multiple" devices)
+#  - current_version (current appVersion)
+#  - current_week (week)
+#  - is_active (were the client_ids active this week or not)
+#  - n_profiles (count of matching client_ids)
+#  - usage_hours (sum of the per-client subsession lengths,
+#        clamped in the[0, MAX_SUBSESSION_LENGTH] range)
+#  - sum_squared_usage_hours (the sum of squares of the usage hours)
+
+MAX_SUBSESSION_LENGTH = 60 * 60 * 48  # 48 hours in seconds.
+
+
+def get_newest_per_client(df):
+    windowSpec = (
+        Window
+        .partitionBy(F.col('client_id'))
+        .orderBy(F.col('timestamp').desc()))
+
+    rownum_by_timestamp = (F.row_number().over(windowSpec))
+    selectable_by_client = df.select(
+        rownum_by_timestamp.alias('row_number'),
+        *[F.col(col) for col in df.columns]
+    )
+    return selectable_by_client.filter(selectable_by_client['row_number'] == 1)
+
+
+def compute_churn_week(df, week_start):
+    """Compute the churn data for this week. Note that it takes 10 days
+    from the end of this period for all the activity to arrive. This data
+    should be from Sunday through Saturday.
+
+    df: DataFrame of the dataset relevant to computing the churn
+    week_start: datestring of this time period"""
+
+    week_start_date = datetime.strptime(week_start, "%Y%m%d")
+    week_end_date = week_start_date + timedelta(6)
+    week_start = fmt(week_start_date)
+    week_end = fmt(week_end_date)
+
+    # Verify that the start date is a Sunday
+    if week_start_date.weekday() != 6:
+        print("Week start date {} is not a Sunday".format(week_start))
```

Should we raise an exception instead of printing a message here? I think we will want to treat a request to run on the wrong day as an ETL error.